### DISCUSSION TITLE
How does one re-run autokerning without changing the kerning classes?

### DISCUSSION THREAD

**Discussion by Tynach** [fontforge/fontforge#5284] [Q&A] [answered]
I have very specific goals for the kerning of my font, and to that end I saved the font to a new file, performed some scaling and stem expansion operations on the copy, and auto-kerned that to get it to generate the correct kerning classes. The resulting kerning itself is completely wrong though, with many pairs being too far apart.. But the classes are all correct, without manual tweaks.

I'm usually pretty happy with the results of auto-kerning anyway, but the incorrect classes caused from *not* doing this was causing me to get bad results, where characters that should be closer together are further apart because of some character in that class not actually belonging to said class.

So... Now I have the font split up into the correct set of classes, but too far apart because the kerning values were calculated from the bloated out version of the font that was only made to get the classes correct. So, how do I get the kerning to re-run with the existing classes intact?

Note: due to the number of classes, I had to also split the kerning up into 4 subtables. I kern the first half of the glyphs with the first half of the glyphs, then I kern the first half with the second half, then the second half with the first half, and finally the second half with the second half. Otherwise, OpenType fonts can't be generated because the total number of classes exceeds 32767 (they seem to consist of 2 bytes, since the actual error complains about it being some high number that I notice to always be over 65535).

It's also worth noting that the settings I use while autokerning are:

- Default Separation: 100
- Min Kern: 100
- ☑ Touching
- ☐ Only kern glyphs closer

I'm not afraid of scripting if that's what it takes, though when I (briefly) looked up how to do it with that route I instead came upon information for how to override FontForge's default ways of determining glyph's optical distances, and stuff like that. Right now though, I just want them 100 units apart exactly, which FontForge can do without custom code.

**Comment by Tynach** [comment, marked as answer]
**TL;DR**: I just had to right-click in the middle of the table view when looking at the subtable I wanted to re-kern, and there in the right-click menu is an option to auto-kern all of them.

The following is the adventure I went on to find this.

---

So, I tried scripting it anyway, and ran into snags. I figured I'd copy the list of glyphs, delete the kerning classes, and then recreate the same kerning classes by having them contain the same glyphs.

The first problem I ran into when actually trying to do this, is that `font.addKerningClass()` *doesn't let you you specify to kern by actual distance*. Non-class-based kerning (`font.autoKern()`) has an optional `touch` parameter that lets you tell it to go by physical distance rather than optical distance, but that parameter is missing from `font.addKerningClass()`.

So, I tried to get around that by setting the `separation` parameter to `0`, and then adding `100` to every kerning pair. This *almost* worked, except that all the values for the 0th kern class (on both sides; so in the class-based kerning dialog, the first column and the first row of the table) were now set to `100`, when they *should* be `0`.

This caused issues with glyphs that I had purposefully not kerned at all. And while I was frustrated and hoping that maybe I could just find *some*thing to set an entire row/column to zero or reset it or just.. *Any*thing, really... I happened to right-click in the middle of the table in the GUI dialog box.

And I saw a right-click menu containing:

> AutoKern Row
> AutoKern Column
> AutoKern All
> ___
> Clear
> Clear All
> Clear Device Table
> Clear All Device Tables

And I felt like a complete moron for not having clicked around in the GUI earlier, because in THIS dialog box there's that 'Touching' checkbox at the top that makes it do EXACTLY what I want.
___

So, anyway, I might still do some scripting and whatnot, so my learning experience wasn't *completely* wasted. In particular, instead of splitting the kerning into 'first half of the font' and 'second half of the font', I can probably do it much more efficiently if I can kern the whole font into one big table, and then split that table into 'first half of the kerning classes' and 'second half of the kerning classes'.

For the curious, here's the almost-working code I ended up with. ***DO NOT USE THIS***, because it sets the first row/column of kerning to `100`, when those should practically *always* be `0`:

```python
fnt = fontforge.activeFont()

kern_subtabs = fnt.getLookupSubtables("'kern' Horizontal Kerning 0")
subtab_kerns = [fnt.getKerningClass(st) for st in kern_subtabs]
subtab_kerns = [(c[0], c[1]) for c in subtab_kerns]

for i, st in enumerate(kern_subtabs):
    fnt.removeLookupSubtable(st)
    fnt.addKerningClass("'kern' Horizontal Kerning 0", st, 0, subtab_kerns[i][0], subtab_kerns[i][1], False, True)

kern_subtabs = fnt.getLookupSubtables("'kern' Horizontal Kerning 0")
subtab_kerns = [fnt.getKerningClass(st) for st in kern_subtabs]
subtab_kerns = [(c[0], c[1], tuple(k + 100 for k in c[2])) for c in subtab_kerns]

for i, st in enumerate(kern_subtabs):
    fnt.removeLookupSubtable(st)
    fnt.addKerningClass("'kern' Horizontal Kerning 0", st, subtab_kerns[i][0], subtab_kerns[i][1], subtab_kerns[i][2])
```
